### PR TITLE
better slug suggestions from library names

### PR DESCRIPTION
**Pull request by 2is10**
@ahsu1230

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 1]
```diff
@@ -163,20 +164,28 @@ object LibrarySlug {
   implicit def format: Format[LibrarySlug] =
     Format(__.read[String].map(LibrarySlug(_)), new Writes[LibrarySlug] { def writes(o: LibrarySlug) = JsString(o.value) })
 
-  val MaxLength = 50
-  val invalidSlugSet = Set("libraries", "connections", "followers", "keeps", "tags", "main", "secret")
+  private val MaxLength = 50
+  private val ReservedSlugs = Set("libraries", "connections", "followers", "keeps", "tags")
+  private val BeforeTruncate = Seq("[^\\w\\s-]|_" -> "", "(\\s|--)+" -> "-", "^-" -> "") map compile
+  private val AfterTruncate = Seq("-$" -> "") map compile
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala:170]: pre-compiling these patterns as an optimization

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 2]
```diff
@@ -163,20 +164,28 @@ object LibrarySlug {
...
   def isReservedSlug(slug: String): Boolean = {
-    invalidSlugSet.contains(slug)
+    ReservedSlugs.contains(slug)
   }
 
   def generateFromName(name: String): String = {
-    name.toLowerCase().replaceAll("[^\\w\\s]|_", "").replaceAll("\\s+", "-").replaceAll("^-", "").take(MaxLength).replaceAll("-$", "") // taken from generateSlug() in  manageLibrary.js
+    // taken from generateSlug() in angular/src/common/util.js
+    val s1 = BeforeTruncate.foldLeft(name.toLowerCase())(replaceAll)
+    val s2 = AfterTruncate.foldLeft(s1.take(MaxLength))(replaceAll)
```
Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala:183]: oo cool, pretty neat way to strip out all the patterns

**Comment by ahsu1230** [comment]
thanks for the change + tests! :+1:

**Comment by 2is10** [comment]
@thassss @mrbrown14 @jaredpetker @DerekBurch 

Here’s the latest code for library slug suggestions on the website and on the server. In addition to auto-appending a `-` to reserved slugs, we also now reduce a series of consecutive `-` down to one. I’d suggest copying these test cases to your own codebases. If you have any other suggestions for improvement or additional test cases, I’ll be happy to incorporate them here.

**Pull request merged**